enumerable.rb

```diff
@@ -0,0 +1,109 @@
+module Enumerable
+  def my_each
+    # Should return the array except block is given
+    return to_enum unless block_given?
+
+    each do |i|
+      yield i
+    end
+  end
+
+  def my_each_with_index
+    # Should return the array except block is given
+    return to_enum unless block_given?
+
+    (0..size - 1).my_each do |i|
+      yield(self[i], i)
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    # an empty array to store our selected element
+    arr = []
+    (0...size - 1).my_each do |i|
+      arr << self[i] if yield(self[i])
+    end
+    arr
+  end
+
+  def my_all?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 unless yield self[i]
+    end
+    count.zero?
+  end
+
+  def my_any?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 if yield self[i]
+    end
+    count != 0
+  end
+
+  def my_none?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 if yield(self[i])
+    end
+    count.zero?
+  end
+
+  def my_count(variable = nil)
+    count = 0
+    (0..size - 1).my_each do |i|
+      if !block_given? && variable.nil?
+        count = size
+      elsif block_given?
+        count += 1 if yield(self[i])
+      else
+        count += 1 unless self[i] != variable
+      end
+    end
+    count
+  end
+
+  def my_map(&proc)
+    arr = []
+    return to_enum unless block_given?
+
+    (0..size - 1).my_each do |i|
+      puts
+      arr << yield(self[i]) if !proc && yield(self[i])
+      arr << yield(self[i]) if proc&.call(self[i])
+    end
+    arr
+  end
```

Comment:
#### my_map 

- [ ] accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

Given 
`my_proc = Proc.new {|num| num < 10 }`

> [18, 22, 5, 6] .my_map(&my_proc) ==  [18, 22, 5, 6] .my_map(&my_proc)  => false false true true

[OPTIONAL]

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend you add it to your my_map

Given 
`my_proc = Proc.new {|num| num > 10 }`

Result
> puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false